### PR TITLE
Rs/logging feature

### PR DESCRIPTION
**Pull request by rsnm2**
@dbogunowicz 

Updated with your comments. Please LMK what you think.

### CODE REVIEW

File: src/content/user-guide/deepsparse-engine/logging.mdx [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+metaTitle: "DeepSparse Logging"
+metaDescription: "System and Data Logging with DeepSparse"
+index: 6000
+---
+
+# DeepSparse Logging
+
+DeepSparse Logging provides you with access to telemetry needed to monitor a deployment. 
+
+There are many types of downstream monitoring tasks that you may want to perform. The difficulty of the tasks varies from relatively easy
```
Comment (dbogunowicz) [inline src/content/user-guide/deepsparse-engine/logging.mdx:12]: what do we mean by "downstream" here?

File: src/content/user-guide/deepsparse-engine/logging.mdx [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+index: 6000
+---
+
+# DeepSparse Logging
+
+DeepSparse Logging provides you with access to telemetry needed to monitor a deployment. 
+
+There are many types of downstream monitoring tasks that you may want to perform. The difficulty of the tasks varies from relatively easy 
+(simple system performance analysis) to challenging (assessing the accuracy of the system in the wild by manually labeling 
+the input data distribution post-facto). Examples include:
```
Comment (dbogunowicz) [inline src/content/user-guide/deepsparse-engine/logging.mdx:14]:
```suggestion
the input data distribution post-factum). Examples include:
```

File: src/content/user-guide/deepsparse-engine/logging.mdx [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+- System performance: what is the latency/throughput of a query?
+- Data quality: is there an issue getting data to my model?
```
Comment (dbogunowicz) [inline src/content/user-guide/deepsparse-engine/logging.mdx:16]:
```suggestion
- Data quality: is the model being fed the expected data distribution?
```

File: src/content/user-guide/deepsparse-engine/logging.mdx [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+production inference pipeline into the logging system of your choice. 
+
+## Metrics 
+DeepSparse Logging provides access to two types of metrics.
+
+### System Logging Metrics 
+
+System Logging Metrics give you access to granular performance metrics for quick and efficient diagnosis of deployment system health.
+
+There is one group of System Logging Metrics currently available: Prediction Latency.  For each inference request, DeepSparse logs the following:
```
Comment (dbogunowicz) [inline src/content/user-guide/deepsparse-engine/logging.mdx:30]:
```suggestion
There is one group of System Logging Metrics currently available: Inference Latency. For each inference request, the DeepSparse Server logs the following:
```

File: src/content/user-guide/deepsparse-engine/logging.mdx [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+1. Pre-processing Time - milliseconds in the pre-processing step
```
Comment (dbogunowicz) [inline src/content/user-guide/deepsparse-engine/logging.mdx:31]: in all cases, the units are seconds, not miliseconds

File: src/content/user-guide/deepsparse-engine/logging.mdx [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+2. Engine Time - milliseconds in the engine forward pass step
+3. Post-processing Time - milliseconds in the post-processing step
+4. Total Time - milliseconds for the end-to-end response time
```
Comment (dbogunowicz) [inline src/content/user-guide/deepsparse-engine/logging.mdx:34]:
```suggestion
4. Total Time - milliseconds for the end-to-end response time (the sum of the first three measurements)
```

File: src/content/user-guide/deepsparse-engine/logging.mdx [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+
+### Data Logging Metrics
+
+Data Logging Metrics give ML teams access to data at each stage of an inference pipeline. 
+This facilitates inspection of the data, understanding of its properties, detecting edge cases and possible data drift.
+
+There are 4 stages in the inference pipeline where Data Logging can occur:
+1. `pipeline_inputs`: raw input passed to the inference pipeline by the user
+2. `engine_inputs`: pre-processed tensors passed to the engine for the forward passed
```
Comment (dbogunowicz) [inline src/content/user-guide/deepsparse-engine/logging.mdx:43]:
```suggestion
2. `engine_inputs`: pre-processed tensors passed to the engine for the forward pass
```

File: src/content/user-guide/deepsparse-engine/logging.mdx [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+3. `engine_outputs`: result of the engine forward pass (e.g. the raw logits)
+4. `pipeline_outputs`: raw output returned to the pipeline caller
```
Comment (dbogunowicz) [inline src/content/user-guide/deepsparse-engine/logging.mdx:45]:
```suggestion
4. `pipeline_outputs`: final output returned to the pipeline caller
```

File: src/content/user-guide/deepsparse-engine/logging.mdx [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
+
+At each stage, you can specify functions to be applied to the data before Logging. Examples functions include the identity function
```
Comment (dbogunowicz) [inline src/content/user-guide/deepsparse-engine/logging.mdx:47]:
```suggestion
At each stage, you can specify functions to be applied to the data before logging. Example functions include the identity function
```

File: src/content/user-guide/deepsparse-engine/logging.mdx [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+
+There are three types of functions that can be applied to target data at each stage:
+1. Built-in functions - pre-written functions provided by DeepSparse ([see list on GitHub](https://github.com/neuralmagic/deepsparse/blob/main/src/deepsparse/loggers/metric_functions/built_ins.py)) 
+2. Framework Functions - functions from `torch` or `numpy` (e.g. `torch.mean`, `np.mean`)
+3. Custom Functions - custom user-provided functions with dynamic imports
+
+## Configuration
+
+A YAML file is used to configure both System and Data Logging.
+- **System Logging** is *enabled* if a `logger` is specified
```
Comment (dbogunowicz) [inline src/content/user-guide/deepsparse-engine/logging.mdx:59]: Right now the default behavior is, that even if the `logger` is not specified, the default logger gets created ( a single PythonLogger)

File: src/content/user-guide/deepsparse-engine/logging.mdx [hunk 11]
```diff
@@ -0,0 +1,319 @@
...
+    - func: path/to/custom.py:bar  
+      frequency: 10000
+      # if no target_loggers, defaults to all
+```
+
+Let's break it down piece by piece:
+
+- `loggers` is a dictionary of the form `{logger_name: {config}}` and identifies and provides the arguments for each logger.
+
+- `data_logging` is a dictionary of the form `{stage.target[idx]: [fn_config]}`.
```
Comment (dbogunowicz) [inline src/content/user-guide/deepsparse-engine/logging.mdx:98]: I don't get this

File: src/content/user-guide/deepsparse-engine/logging.mdx [hunk 12]
```diff
@@ -0,0 +1,319 @@
...
+properties, an optional `target` can be used to unpack the data. If the `target` is a dict or a list, 
+you can access via slicing, indexing, or dict access, for example `stage.target[0:2][0, :-1]['key_3'][5]`.
+
+- `[fn_config]` is list of function configurations of the form `{func: name, frequency: freq, target_loggers: [logger_names]}`. 
+It specifies which the function, the logger, and the frequency of the logging.
+
+<details>
+  <summary>Click for a tangible example for an image classification endpoint.</summary>
+
+```yaml
```
Comment (dbogunowicz) [inline src/content/user-guide/deepsparse-engine/logging.mdx:111]: Should we add those doc examples also as items for the QA to test?

File: src/content/user-guide/deepsparse-engine/logging.mdx [hunk 13]
```diff
@@ -0,0 +1,319 @@
...
+          target_loggers:
+            - python
+        - func: np.shape        # framework function
+          frequency: 1
+          target_loggers:
+            - python
+```
+
+This configuration does the following data logging at each respective stage of the pipeline:
+- System logging is enabled by default and logs to Prometheus
```
Comment (dbogunowicz) [inline src/content/user-guide/deepsparse-engine/logging.mdx:151]: We are explicitly enabling both Prometheus and Python logger no?

File: src/content/user-guide/deepsparse-engine/logging.mdx [hunk 14]
```diff
@@ -0,0 +1,319 @@
...
+- Logs the shape of the input batch provided by the user to stdout
```
Comment (dbogunowicz) [inline src/content/user-guide/deepsparse-engine/logging.mdx:152]: I like that breakdown.

File: src/content/user-guide/deepsparse-engine/logging.mdx [hunk 15]
```diff
@@ -0,0 +1,319 @@
...
+accepts no arguments and is configured with the following:
+
+```yaml
+loggers:
+  python:
+```
+
+### Prometheus Logger
+
+DeepSparse is integrated with Prometheus. The Prometheus Logger accepts an optional `port` property
```
Comment (dbogunowicz) [inline src/content/user-guide/deepsparse-engine/logging.mdx:176]: Not only port,  there are more arguments that PrometheusLogger can take.

File: src/content/user-guide/deepsparse-engine/logging.mdx [hunk 16]
```diff
@@ -0,0 +1,319 @@
...
+    name: image_classification_pipeline
+    data_logging:
+      pipeline_inputs.images:   # applies to the images (of the form stage.target:)
+        - func: np.shape        # framework function
+          frequency: 1
+          target_loggers:
+            - python          
+
+      pipeline_inputs.images[0]:          # applies to the first image (of the form stage.target[idx]:)
+        - func: mean_pixels_per_channel   # built-in function
```
Comment (dbogunowicz) [inline src/content/user-guide/deepsparse-engine/logging.mdx:130]: This fails as expected, the function outputs a triple and prometheus does not log triples, only ints/floats

File: src/content/user-guide/deepsparse-engine/logging.mdx [hunk 17]
```diff
@@ -0,0 +1,319 @@
...
+    model: zoo:cv/classification/resnet_v1-50/pytorch/sparseml/imagenet/pruned95_quant-none
+    name: image_classification_pipeline
+    data_logging:
+      pipeline_inputs.images:
+        - func:
+          frequency: 1
+          target_loggers:
+            - python
+
+      pipeline_inputs.images[0]
```
Comment (dbogunowicz) [inline src/content/user-guide/deepsparse-engine/logging.mdx:265]:
```suggestion
      pipeline_inputs.images[0]:
```

File: src/content/user-guide/deepsparse-engine/logging.mdx [hunk 18]
```diff
@@ -0,0 +1,319 @@
...
+      
+      pipeline_outputs.scores[0]:
+        - func: identity
+          frequency: 1
+      
+      pipeline_outputs.labels[0]:
+        - func: identity
+          frequency: 1
+
+  - task: sentiment/analysis
```
Comment (dbogunowicz) [inline src/content/user-guide/deepsparse-engine/logging.mdx:287]:
```suggestion
  - task: sentiment_analysis
```

File: src/content/user-guide/deepsparse-engine/logging.mdx [hunk 19]
```diff
@@ -0,0 +1,319 @@
...
+          frequency: 1
+          target_loggers:
+            - python
+
+      pipeline_inputs.images[0]
+        - func: max_pixels_per_channel
+          frequency: 1
+          target_loggers:
+            - prometheus
+        - func: mean_pixels_per_channel
```
Comment (dbogunowicz) [inline src/content/user-guide/deepsparse-engine/logging.mdx:270]: will fail with prometheus

File: src/content/user-guide/deepsparse-engine/logging.mdx [hunk 20]
```diff
@@ -0,0 +1,319 @@
...
+    name: image_classification_pipeline
+    data_logging:
+      pipeline_inputs.images:
+        - func:
+          frequency: 1
+          target_loggers:
+            - python
+
+      pipeline_inputs.images[0]
+        - func: max_pixels_per_channel
```
Comment (dbogunowicz) [inline src/content/user-guide/deepsparse-engine/logging.mdx:266]: will fail with prometheus

File: src/content/user-guide/deepsparse-engine/logging.mdx (hunk 3 above)
Comment (rsnm2) [reply src/content/user-guide/deepsparse-engine/logging.mdx:16]: Data quality is about detecting errors in the data pipeline. For instance, detecting if the camera is broken and as a result you are getting all 0's. Vs data distribution is slightly different and we have a separate bullet for that
